[lldb] progressive progress reporting for darwin kernel/firmware

jasonmolenda opened the pull request:
When doing firmware/kernel debugging, it is frequent that binaries and debug info need to be retrieved / downloaded, and the lack of progress reports made for a poor experience, with lldb seemingly hung while downloading things over the network. This PR adds progress reports to the critical sites for these use cases.

jasonmolenda committed:
[lldb] progressive progress reporting for darwin kernel/firmware

When doing firmware/kernel debugging, it is frequent that
binaries and debug info need to be retrieved / downloaded,
and the lack of progress reports made for a poor experience,
with lldb seemingly hung while downloading things over the
network. This PR adds progress reports to the critical sites
for these use cases.

jasonmolenda commented:
The one thing I was uncertain of is when I load kexts -- I have a known number of kexts that I will be attempting to load -- and I wasn't sure if I should log them individually as they happen (what I did), or if I should initialize a progress category with the number of kexts that lldb will be attempting to load, so the progress of the job overall can be visualized.

medismailben commented:
> The one thing I was uncertain of is when I load kexts -- I have a known number of kexts that I will be attempting to load -- and I wasn't sure if I should log them individually as they happen (what I did), or if I should initialize a progress category with the number of kexts that lldb will be attempting to load, so the progress of the job overall can be visualized.

If we know ahead of time how many kext will be loaded, I think it would be better to provide that number to the progress object constructor, and pass the name of the kext (and other relevant information) in the [`Increment` upgraded_details string argument](https://github.com/llvm/llvm-project/blob/main/lldb/include/lldb/Core/Progress.h#L104). Otherwise this is great! LGTM!

jasonmolenda committed:
locate all progress reporting for kexts and kernels in one method,
LoadImageUsingMemoryModule.  Remove the other progress reporting
from ParseKextSummaries.  Add 0x prefix manually before PutHex64.

In LoadImageUsingMemoryModule add the kext/kernel name to the
ModuleSpec so it will be printed by the "Loading symbol file for
..." progress report in the generic DebugSymbols printer.  Remove
the logic that was only calling the Platform GetSharedModule with
the kext/kernel filename if it was PlatformDarwinKernel.  First,
it should always be PlatformDarwinKernel already, but even if it's
not, the kext bundle ID or placeholder "mach_kernel" will not match
a real file, the restriction to only use the name for
PlatformDarwinKernel was unnecessary.

If I have a correct UUID for the kext/kernel, don't add a possibly
incorrect ArchSpec to the ModuleSpec.

jasonmolenda commented:
Ah no, I misunderstood.  The Increment method of Progress is intended for use where you have one progress status display that takes multiple steps to complete.  It's a separate mechanism from "finding dSYM for file x" etc.  I can't use an Increment style progress reporting in this PR.

clayborg commented on lldb/source/Core/DynamicLoader.cpp:
```diff
@@ -195,20 +196,40 @@ ModuleSP DynamicLoader::LoadBinaryWithUUIDAndAddress(
+  }
+
   if (!uuid.IsValid() && !value_is_offset) {
+    Progress progress_memread("Reading load commands from memory",
```
I would change this string to "Reading object file from memory" instead of "Reading load commands from memory". This function is currently only used by Mach-O core files and the GDB remote stub, but this could be used for loading ELF files from memory as well. 

Maybe this Progress dialog should go into `ReadUnnamedMemoryModule` so anyone that calls this function gets progress events when loading an object file from memory?

jasonmolenda replied:
Good suggestion.  I think I'll conditionalize this progress with a `if (!process->GetCoreFile())` -- reading a binary out of a local corefile will be very fast.  Over a live gdb-remote connection, not so much.

clayborg commented on lldb/source/Core/DynamicLoader.cpp:
```diff
@@ -195,20 +196,40 @@ ModuleSP DynamicLoader::LoadBinaryWithUUIDAndAddress(
-    module_spec.GetFileSpec() = name_filespec;
 
   if (uuid.IsValid()) {
+    Progress progress("Locating external symbol file",
+                      prog_str.GetString().str());
+
     // Has lldb already seen a module with this UUID?
+    // Or have external lookup enabled in DebugSymbols on macOS.
```
Should this go into `ModuleList::GetSharedModule()` or at some function that is called by `ModuleList::GetSharedModule()`? If the `ModuleList::GetSharedModule()` function is really quick, I hate to add really spammy progress notifications that aren't useful. If there is code down in `ModuleList::GetSharedModule()` that locates an external symbol file, then the progress should go in that function that is known to take some amount of time, but not here in a general area IMHO

jasonmolenda replied:
It's a little tricky because we do call ModuleList::GetSharedModule which will return a module if it's already in the global module cache, and it will call into the DebugSymbols framework on macOS, where it might call an external program to do a slow copy of a binary to the local computer.   But after that, it then goes on to call `LocateExecutableSymbolFile` `LocateExecutableObjectFile` looking in known local filesystem locations.  If that fails, then it calls `DownloadObjectAndSymbolFile (force_symbol_search=true)` which can call out to an external program to do a slow copy of a binary to the local computer.

I wanted to log one message to cover possibly all of these being run.

clayborg replied:
My only concern here is this will spam the progress notifications and also be a bit misleading. The title says "Locating external symbol file" when what this is actually doing is loading the module from a module_spec. It _might_ end up locating an external symbol file. We know reading modules from memory is slow, so fine to make a progress for this, but It would be better to actually put this in the `DownloadObjectAndSymbolFile()` with this label, or change the label to something more like "Loading module" with the detail being the `prog_str`

clayborg commented on lldb/source/Plugins/DynamicLoader/Darwin-Kernel/DynamicLoaderDarwinKernel.cpp:
```diff
@@ -757,11 +758,32 @@ bool DynamicLoaderDarwinKernel::KextImageInfo::LoadImageUsingMemoryModule(
       ModuleSpec module_spec;
       module_spec.GetUUID() = m_uuid;
-      module_spec.GetArchitecture() = target.GetArchitecture();
+      if (!m_uuid.IsValid())
+        module_spec.GetArchitecture() = target.GetArchitecture();
```
Did you mean to check if the UUID is valid in the if statement and then set the architecture?

jasonmolenda replied:
If I have a UUID, it's authoritative, whereas the ArchSpec might be heuristically determined.  I don't like setting both in a ModuleSpec if the UUID is valid, it it noramlly fine but it's a little footgun waiting for some unusual combination where the heuristically determined ArchSpec is not quite the same ("compatible") with the arch of the UUID specified.

jasonmolenda replied:
I know I shouldn't be making changes like this at the same time as adding progress status logging, but it irked me when I saw it.

jasonmolenda replied:
just to be clear, it's almost always the "environment" or "os" part of the triple, which is nearly an entire fiction with firmware style debugging, that is the problem.  One binary will say "hey I'm iOS" and another binary that needs to be loaded also is like "I'm something else" and lldb will reject the module load even though the UUIDs match.

jasonmolenda committed:
Changes to address Greg's feedback.

Move the logging about reading a binary from memory from
DynamicLoader::LoadBinaryWithUUIDAndAddress into
Process::ReadModuleFromMemory.  Rename the progress title
to not be mach-o specific.  Only issue the progress update
if this is a non-corefile Process; lldb will read the
binary out of the corefile so fast that logging is not
useful.

jasonmolenda commented:
Thanks for the helpful feedback @clayborg pushed an update.

jasonmolenda committed:
Create a Progress object for kext loading, incremental updates

In `ParseKextSummaries()` I know how many kexts will be loaded,
create a Progress object there with that number, pass it in to
`LoadImageUsingMemoryModule()`, and issue progressive updates if
the caller passed in an existing Progress object.  Else do a one-off
progress update.  This give us a 1-out-of-n number for the updates,
e.g.

[30/180] Loading kext: com.apple.Libm 5D39F799-ADEA-3CAE-8A7B-830CE3EAEF72 at 0xfffffe002aa6b98..

Thanks to Chelsea for pointing out how I could use the API more
correctly.

clayborg commented on lldb/source/Target/Process.cpp:
```diff
@@ -2545,6 +2546,11 @@ ModuleSP Process::ReadModuleFromMemory(const FileSpec &file_spec,
   if (module_sp) {
     Status error;
+    std::unique_ptr<Progress> progress_up;
+    if (!GetCoreFile())
```
Might be better to use the bool accessor function:
```
if (IsLiveDebugSession())
```

jasonmolenda replied:
The name is a little misleading, `IsLiveDebugSession` is only false for `PostMortemProcess`.  Maybe we should have a `Process::IsCorefileProcess` or something, or I could put a comment there about what I'm doing.

clayborg replied:
`PostMortemProcess` is what all of the core file processes (elf, mach-o and minidump) inherit from. That way all core file processes don't forget to override this function. I agree the name isn't the most clear.

jasonmolenda replied:
Oh whoops, my bad on that one, thanks.  I'll change to use this instead, it's clearer.

jasonmolenda committed:
Add a comment why I check for corefiles & clarify msg

jasonmolenda committed:
ws fix

jasonmolenda committed:
Clarify my corefile check in ReadModuleFromMemory

Greg's suggestion is better.

jasonmolenda committed:
format nit
